src/noit_metric_tag_search.c

```diff
@@ -65,21 +75,58 @@ static tls_state_t *tls_state_alloc(void);
 static pcre_jit_stack *tls_state_get_pcre_jit_stack(void *arg);
 static pthread_key_t tls_state_key;
 
-static re_matcher_t *
-re_matcher_alloc(char *pattern)
+static bool re_matcher_compile(re_matcher_t *rem)
 {
-  const char *error;
-  re_matcher_t *rem = malloc(sizeof(*rem));
-
-  rem->re_str = pattern;
-  rem->re = pcre_compile(pattern, 0, &error, &rem->error_offset, NULL);
-
+  pcre *re = NULL;
+  if((re = ck_pr_load_ptr(&rem->re)) == NULL && rem->error_offset < 0) {
+    const char *error;
+    re = pcre_compile(rem->re_str, 0, &error, &rem->error_offset, NULL);
+    if(!ck_pr_cas_ptr(&rem->re, NULL, re)) {
+      pcre_free(re);
+      re = ck_pr_load_ptr(&rem->re);
+    }
+  }
+  return re != NULL;
+}
+static inline bool re_matcher_study(re_matcher_t *rem)
+{
+  pcre_extra *re_e = ck_pr_load_ptr(&rem->re_e);
```

Comment:
no need for atomics, the cache is thread local, it has to be for the jit stack to work properly too

Reply:
It doesn't need to be thread local for the jit stack to work, that's why we pass in a function to create the stack and not just pass in memory.  This tls jit stack factory is used elsewhere as well.

Reply:
How can this possibly work?  Once the stack is assigned to a regex, it's still the same shared stack for all the assignments that were done on this thread.

If the regex is moved with its stack to another thread, then it will be used by multiple threads at the same time.

The study should only be done when requesting a regex from this API, at the very least this would limit the atomics to only the execution counter

Reply:
b/c the stack isn't assigned to the regex.  The function to get the stack is called as needed.  And it returns a thread local stack each time.

Reply:
checking the source of pcre_assign_jit_stack to see if the function is called right away or when executing and what it's assigned to

Reply:
> b/c the stack isn't assigned to the regex. The function to get the stack is called as needed. And it returns a thread local stack each time.

OK then, but this should still be a single atomic for the count.  We get it from the cache and we check its execution counter and swap it in